### DISCUSSION TITLE
【IOS】How to generate a video using copyFrame api in ios device.

### DISCUSSION THREAD

**Discussion by dandingol03** [google/filament#4897] [Q&A] [answered]
Hi. I am new to ios, and i don't know how to generate a video which contains the whole filament rendered stuff. I can use MediaRecorder to generate a surface in android device, and this surface can be feeded to the `swapchain`. However i don't know how to generate the surface in ios device.  Is anyone who have an idea?

**Comment by bejado** [comment, marked as answer]
This is possible with Filament's Metal backend. Instead of passing a `CAMetalLayer` to `createSwapChain`, for each frame pass in a `CVPixelBuffer` and set the `CONFIG_APPLE_CVPIXELBUFFER` flag. You can use Apple's [AVFoundation API](https://developer.apple.com/documentation/avfoundation/media_assets_and_metadata/sample-level_reading_and_writing?language=objc) to encode the `CVPixelBuffer` frames into a video file.

**Reply by dandingol03** [reply]
Thanks. This help me a lot

**Reply by dandingol03** [reply]
@bejado So if i want to generate a video, i just need to create **one** `CVPixelBuffer`. Rendering filament's stuff into a `SwapChain` which is created by the`CVPixelBuffer` after every `Render::render`'s call, and encode **the only one** `CVPixelBuffer` to the video file. Am I right?

**Reply by bejado** [reply]
I believe you'll want a `CVPixelBuffer` for each frame. If I remember correctly, once you submit a buffer for encoding you're not allowed to modify it until AVFoundation has finished processing it. In the past I've had success creating 3-4 `CVPixelBuffer`s and cycling through them round-robin style. That gives the device enough time to encode the frame before reusing the buffer.

**Comment by ritik619** [comment]
Hello @dandingol03 
May i know if you were able to do so in Android ?
If so can you help me out ,I'm stuck

**Reply by dandingol03** [reply]
@ritik619   Yes, i make it work in android too. Create the capture swapchain from the surface which are generated by `MediaRecorder`, and use `copyFrame` api in every draw call to copy the frame to your capture swapchain.
```
    if (renderer->beginFrame(params->recordSwapChain, frameTimeNanos)) {
      renderer->render(params->recordView_);
      renderer->copyFrame(params->captureSwapChain, ...);            
    }
    renderer->endFrame();
```
